zokrates_stdlib/stdlib/utils/casts/field_to_u64.zok

```diff
@@ -1,5 +1,9 @@
 from "EMBED" import unpack, u64_from_bits
 
-def main(field i) -> u64:
-    bool[64] bits = unpack(i)
+const field U64_MAX = 18446744073709551615
+
+// Condition: field value is asserted to be in range [0, U64_MAX]
+def main(field input) -> u64:
+    assert(input <= U64_MAX)
```

Comment:
The behavior would be the same as in most other programing languages if instead of asserting `input < 2^64`, the return value were set to `input mod 2^64`.

Reply:
I can see that argument. Can you elaborate on your main use case for the truncating behavior? Is it when packing a bunch of arguments into one field outside the circuit, then unpacking them inside knowing they are all in range? Or do you actually have cases where truncation happens?

Reply:
Yes, I have a use case where truncation happens. When I implemented long integer multiplication, it would've been useful to represent a long integer as an array of u64s (I switched to u32s by now because of the problems). However, I of course cannot multiply two u64s and then access the result as multiplication would then happen mod 2^64. I have to use field elements that each represent a 64 bit block of a long integer. If I multiply them, the entire result is there. However, I need to get back to u64, or rather two of them: one for the result of the current position, one for the carry. So the idea was to just stuff the field element into a u64, call that the current element, and subtract that from the multiplication result, divide by 2^64, convert to u64, and call that the carry. However, the compiler crashed as soon as the multiplication result was so big that the carry was actually needed.